Remove static metaFields list and get version-specific values from core

cwperks opened the pull request:
### Description

Remove static metaFields list and get version-specific values from core.

The new list obtained from core is: `[_ignored, _routing, _seq_no, _doc_count, _index, _nested_path, _data_stream_timestamp, _field_names, _source, _id, _version, _primary_term]`

This doesn't match the hardcoded list because it removes many of the deprecated fields from [META_FIELDS_BEFORE_7DOT8](https://github.com/opensearch-project/OpenSearch/blob/8883e62a8f3d862739250cff1eb8d3d18680cfa2/server/src/main/java/org/opensearch/index/mapper/MapperService.java#L204-L208) 

When using FLS Includes rules, this list is used as the starting point for all metaFields to include by default.

* Category

Bug fix

### Issues Resolved

- https://github.com/opensearch-project/security/issues/4349

### Check List
- [ ] New functionality includes testing
- [ ] New functionality has been documented
- [ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

cwperks commented on src/main/java/org/opensearch/security/configuration/SecurityFlsDlsIndexSearcherWrapper.java:
```diff
@@ -75,6 +56,7 @@ public SecurityFlsDlsIndexSearcherWrapper(
         final Salt salt
     ) {
         super(indexService, settings, adminDNs, evaluator);
+        metaFields = indexService.mapperService().getMetadataFields();
```
I updated the PR description so you can see what the list obtained from core contains. The lists are different, but the values being removed are marked as meta fields before ES 7.8. 

It doesn't break any existing test so there is that assurance. If anything does break, it means that our tests for FLS with includes rules are not extensive enough.

DarshitChanpura replied:
okay, i think we should add a test that breaks this behavior to validate what we expect. Not sure if one exists already.

RyanL1997 reviewed (approved):
Is this a `backport 2.x` ?

cwperks commented:
> Is this a backport 2.x ?

Yes it is. I added the label.

peternied reviewed (dismissed):
Lets add a test case that verifies that all previous meta items are still covered so we aren't leaking any sensitive data.

I'm curious on your thoughts, how do you think we can build confidence around meta field exposure?  I'm not sure we have strong coverage right now.

cwperks commented:
@peternied See the PR description. The new list is different and removes the legacy fields: `"_size", "_timestamp", "_ttl", "_type"` (Fields from this [list](https://github.com/opensearch-project/OpenSearch/blob/8883e62a8f3d862739250cff1eb8d3d18680cfa2/server/src/main/java/org/opensearch/index/mapper/MapperService.java#L205-L208)).

On the backport to 2.x, I plan to keep those fields because they are still referenced here: https://github.com/opensearch-project/OpenSearch/blob/2.x/server/src/main/java/org/opensearch/search/SearchHit.java#L188-L201

This datastructure is used as the default `includesSet` for FLS when using includes rules. Effectively, that means that these are the fields that are explicitly included without having to list them out in the FLS portion of a role. If keys are removed from this list then they will not be returned in the result set when performing a search query when a user is mapped to a role containing FLS includes rules.

peternied commented:
> The new list is different and removes the legacy fields: "_size", "_timestamp", "_ttl", "_type" (Fields from this [list](https://github.com/opensearch-project/OpenSearch/blob/8883e62a8f3d862739250cff1eb8d3d18680cfa2/server/src/main/java/org/opensearch/index/mapper/MapperService.java#L205-L208)).

@cwperks Thanks - that's a great reason to tweak the test case coverage based on fields listed above.  Is there a reason you think we shouldn't add a test case that would be sure that say `_source` is specifically filtered as we expect?

cwperks commented:
@peternied I will flip this to Draft while a test case is worked on and to understand the legacy fields in greater detail.

cwperks commented on src/main/java/org/opensearch/security/configuration/SecurityFlsDlsIndexSearcherWrapper.java:
```diff
@@ -38,24 +40,9 @@
-        "_type"
+    private final Set<String> metaFields;
+    public static final Set<String> META_FIELDS_BEFORE_7DOT8 = Collections.unmodifiableSet(
+        new HashSet<>(Arrays.asList("_size", "_timestamp", "_ttl", "_type"))
```
@peternied Quick update: "_ttl" and "_timestamp" seem to have been removed in ES 6.

The `TypeFieldMapper` was removed in OS 2.0: https://github.com/opensearch-project/OpenSearch/pull/3196

The "_size" FieldMapper comes from `mapper-size` plugin: https://github.com/opensearch-project/OpenSearch/blob/main/plugins/mapper-size/src/main/java/org/opensearch/index/mapper/size/SizeFieldMapper.java

I'm not sure if that should be considered built-in and in this includesSet by default.

peternied replied:
@cwperks Thanks for looking into this and charting out the history of these fields.

If `indexService.mapperService().getMetadataFields()` is getting metadata fields from the current index then it seems like the source of truth and we don't need a backup list, can you look into this?

cwperks replied:
Yes, looking into it now. To recap: this list of metaFields is used by the security plugin for FLS when a role has FLS Include rules (as opposed to excludes). The idea of this list, is that metaFields are included by default in addition to fields that a user explicitly allows when using FLS includes.

The fields returned from `getMetadataFields` are driven by [IndicesModule.getMetadataMappers](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/IndicesModule.java#L212-L242) which includes the built-in metadata mappers + mapper plugins (such as the mapper-size plugin).

I am going to create a localDistro of OpenSearch, install the mapper-size plugin and verify that "_size" is returned from `indexService.mapperService().getMetadataFields()`.

If that's the case, then I feel comfortable flipping this from draft and explicitly removing `"_timestamp", "_ttl", "_type"` which are legacy mappers "_timestamp" and "_ttl" were removed in ES 6 and "_type" was changed to "_nested_path" in OS 2.

cwperks replied:
The documentation on the mapper-size plugin is lacking. To setup the mapper-size plugin you can create a local distribution of core and use `./bin/opensearch-plugin install mapper-size`.

Alternatively, you can create a zip and install it using:

```
./gradlew :plugins:mapper-size:assemble
./bin/opensearch-plugin install file:$OPENSEARCH_HOME/plugins/mapper-size/build/distributions/mapper-size-3.0.0-SNAPSHOT.zip
```

After you startup a cluster, you can create an index with the size mapping enabled, index a document and query like this:

```
curl -XPUT http://localhost:9200/example_index -H "Content-Type: application/json" -d '{
  "mappings": {
    "_size": {
      "enabled": true
    },
    "properties": {
      "name": {
        "type": "text"
      },
      "age": {
        "type": "integer"
      }
    }
  }
}'

curl -XPOST http://localhost:9200/example_index/_doc -H "Content-Type: application/json" -d '{
  "name": "John Doe",
  "age": 30
}'

curl -XGET http://localhost:9200/example_index/_search -H "Content-Type: application/json" -d '{
  "query": {
    "match_all": {}
  },
  "stored_fields": ["_size", "_source"]
}'

Query Results:

{"took":2,"timed_out":false,"_shards":{"total":1,"successful":1,"skipped":0,"failed":0},"hits":{"total":{"value":1,"relation":"eq"},"max_score":1.0,"hits":[{"_index":"example_index","_id":"Pctw0I8BLto8I5f_NLKK","_score":1.0,"_size":37,"_source":{
  "name": "John Doe",
  "age": 30
}}]}}
```

cwperks commented:
@peternied Setting this back to in review. I can't find a way to assert that the list after this change is a superset of the list before this change, but I also don't think that would be a good test case either. For the fields that have been removed like "_timestamp", "_ttl" and "_type", I have added them back in to keep the list the same as before. I have also added "_size", but if someone installed the mapper-size plugin then "_size" will be added twice (its fine since its a set). 

The list of built-in metadata fields can be found here: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/IndicesModule.java#L190-L210

I plan to open a PR on the documentation-website to document the mapper-size plugin. 

Let me know what you think.

peternied reviewed (commented):
@cwperks Thanks for the update on this PR, I'm still on the fence about merging because there isn't any associated test cases, is there an existing test that will catch issues if this list starts behaving strange?

cwperks commented:
@peternied I just pushed a new commit that includes a test.

To write a test, I decided to leverage the mapper-size plugin from core which is not installed by default. If this logic is working correctly, "_size" should be included in the list [given to the security plugin from core](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/IndicesModule.java#L212-L237) since the MapperSize plugin is installed. I updated the integration test for FlsAndFieldMaskingTests to add a test where it creates an index and enables "_size" on the index. The test does a search request against the index with a user that's only allowed to see a single field "stars". The test verifies that the user can see "stars", but not other "fields" indicating that FLS Includes is working as intended. The test also verifies that "_size" is returned in the response indicating that core gave the security plugin that field.

If you explicitly remove "_size" from the metaFields tracked by SecurityFlsDlsIndexSearcherWrapper the test will fail because it would be filtered out by the FLS restrictions. 

FYI "_size" will not be included in the list by default like earlier, however, it will be included for clusters where the MapperSize plugin is installed the this automated test proves it.

peternied reviewed (approved):
@cwperks Nicely done thanks!
